src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -36,43 +33,53 @@ internal sealed class FileChangeWatcher
         /// the end of the queue.
         /// </summary>
-        private Task<IVsAsyncFileChangeEx> _taskQueue;
-        private static readonly Func<Task<IVsAsyncFileChangeEx>, object, Task<IVsAsyncFileChangeEx>> _executeActionDelegate =
-            async (precedingTask, state) =>
-            {
-                var action = (Func<IVsAsyncFileChangeEx, Task>)state;
-                await action(precedingTask.Result).ConfigureAwait(false);
-                return precedingTask.Result;
-            };
+        private readonly AsyncQueue<Func<IVsAsyncFileChangeEx, CancellationToken, Task>> _taskQueue = new();
+        private readonly Task _worker;
 
-        public FileChangeWatcher(Task<IVsAsyncFileChangeEx> fileChangeService)
-            => _taskQueue = fileChangeService;
+        public FileChangeWatcher(Func<Task<IVsAsyncFileChangeEx>> fileChangeService)
+        {
+            _worker = Task.Run(() => RunAsync(fileChangeService, CancellationToken.None));
+        }
 
-        private void EnqueueWork(Func<IVsAsyncFileChangeEx, Task> action)
+        private async Task RunAsync(Func<Task<IVsAsyncFileChangeEx>> getFileChangeService, CancellationToken cancellationToken)
         {
-            lock (_taskQueueGate)
+            var fileChangeService = await getFileChangeService().ConfigureAwait(false);
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var item = await DequeueAsync(_taskQueue, cancellationToken).ConfigureAwait(false);
+                try
+                {
+                    await item(fileChangeService, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception ex) when (FatalError.ReportAndCatchUnlessCanceled(ex, cancellationToken))
+                {
+                }
+            }
+
+            // Local functions
+            static async ValueTask<Func<IVsAsyncFileChangeEx, CancellationToken, Task>> DequeueAsync(AsyncQueue<Func<IVsAsyncFileChangeEx, CancellationToken, Task>> taskQueue, CancellationToken cancellationToken)
             {
-                _taskQueue = _taskQueue.ContinueWith(
-                    _executeActionDelegate,
-                    action,
-                    CancellationToken.None,
-                    TaskContinuationOptions.None,
-                    TaskScheduler.Default).Unwrap();
+                if (taskQueue.TryDequeue(out var value))
+                    return value;
+
+                return await taskQueue.DequeueAsync(cancellationToken).ConfigureAwait(false);
             }
         }
 
+        private void EnqueueWork(Func<IVsAsyncFileChangeEx, CancellationToken, Task> action)
+        {
+            _taskQueue.Enqueue(action);
+        }
+
         // TODO: remove this when there is a mechanism for a caller of EnqueueWatchingFile
         // to explicitly wait on that being complete.
```

Comment:
This would be trivially easy if using `ReentrantSemaphore` because each submitter to the 'queue' implicitly gets a `Task` back immediately.